File: scripts/check-duplicate-types.js
#!/usr/bin/env node
/**
 * Check for duplicate type definitions between @/shared/api/types.ts
 * and @/generated/orval/models to prevent redundancy.
 *
 * Usage: node scripts/check-duplicate-types.js
 */

const fs = require('fs');
const path = require('path');

const projectRoot = path.resolve(__dirname, '..');
const webRoot = path.join(projectRoot, 'apps/web');
const srcRoot = path.join(webRoot, 'src');

/**
 * Extract exported type/interface names from a TypeScript file.
 */
function extractExportedTypes(filePath) {
  if (!fs.existsSync(filePath)) {
    return [];
  }

  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split('\n');
  const types = [];

  // Match: export interface Foo {, export type Foo =, export enum Foo {, etc.
  const exportRegex = /export\s+(interface|type|enum)\s+(\w+)/;

  lines.forEach((line, index) => {
    const match = line.match(exportRegex);
    if (match) {
      const name = match[2];
      types.push({
        name,
        file: filePath,
        line: index + 1,
      });
    }
  });

  return types;
}

/**
 * Recursively get all .ts files from a directory.
 */
function getAllTypeScriptFiles(dir) {
  const files = [];

  function traverse(currentDir) {
    try {
      const entries = fs.readdirSync(currentDir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(currentDir, entry.name);
        if (entry.isDirectory()) {
          traverse(fullPath);
        } else if (entry.isFile() && entry.name.endsWith('.ts')) {
          files.push(fullPath);
        }
      }
    } catch (err) {
      // Skip directories we can't read
    }
  }

  traverse(dir);
  return files;
}

/**
 * Get relative path from web root for display.
 */
function getRelativePath(filePath) {
  return path.relative(webRoot, filePath);
}

/**
 * Main check function.
 */
function checkDuplicateTypes() {
  const handwrittenTypesPath = path.join(srcRoot, 'shared/api/types.ts');
  const orvalModelsDir = path.join(srcRoot, 'generated/orval/models');

  console.log('📋 Checking for duplicate type definitions...\n');

  // Extract types from handwritten types file
  const handwrittenTypes = extractExportedTypes(handwrittenTypesPath);
  console.log(`Found ${handwrittenTypes.length} types in @/shared/api/types.ts`);

  // Get all orval model files
  const orvalFiles = getAllTypeScriptFiles(orvalModelsDir);
  console.log(`Scanning ${orvalFiles.length} files in @/generated/orval/models\n`);

  // Extract types from all orval models
  const orvalTypeMap = new Map();
  for (const file of orvalFiles) {
    const types = extractExportedTypes(file);
    for (const type of types) {
      if (!orvalTypeMap.has(type.name)) {
        orvalTypeMap.set(type.name, []);
      }
      orvalTypeMap.get(type.name).push(type);
    }
  }

  // Check for duplicates
  const duplicates = [];

  for (const handWritten of handwrittenTypes) {
    const orvalMatches = orvalTypeMap.get(handWritten.name);
    if (orvalMatches && orvalMatches.length > 0) {
      duplicates.push({
        name: handWritten.name,
        handwritten: handWritten,
        orval: orvalMatches,
      });
    }
  }

  if (duplicates.length === 0) {
    console.log('✅ No duplicate types found. All handwritten types are unique.\n');
    return 0;
  }

  console.log(`❌ Found ${duplicates.length} duplicate type(s):\n`);

  for (const dup of duplicates) {
    console.log(`  📌 ${dup.name}`);
    console.log(
      `     Handwritten: ${getRelativePath(dup.handwritten.file)}:${dup.handwritten.line}`
    );
    for (const orvalType of dup.orval) {
      console.log(
        `     Generated:   ${getRelativePath(orvalType.file)}:${orvalType.line}`
      );
    }
    console.log();
  }

  console.log(
    '⚠️  These types should be removed from @/shared/api/types.ts'
  );
  console.log('   and imported directly from @/generated/orval/models instead.\n');

  return 1;
}

// Run
const exitCode = checkDuplicateTypes();
process.exit(exitCode);

File: scripts/check-duplicate-types.py
#!/usr/bin/env python3
"""Check for duplicate types between shared/api/types{,ui-helpers} and generated/orval/models.

Reports any type defined in shared/api/types.ts or shared/api/ui-helpers.ts that
has a corresponding (exact or suffix-matched) type in generated/orval/models.
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
WEB_SRC = ROOT / "apps" / "web" / "src"

TYPES_FILE = WEB_SRC / "shared" / "api" / "types.ts"
UI_HELPERS_FILE = WEB_SRC / "shared" / "api" / "ui-helpers.ts"
ORVAL_INDEX = WEB_SRC / "generated" / "orval" / "models" / "index.ts"

# Types in types.ts that are legitimately frontend-only (not API DTOs)
FRONTEND_ONLY_TYPES: set[str] = {
    "TaskType",
    "DatasetType",
    "ModelFramework",
    "ModelFormat",
    "OrgRole",
    "ViewRowV1",
    "ViewPaginatedResponse",
    "PaginatedResponse",
}


def filename_to_type(name: str) -> str:
    """Convert a filename base to its exported PascalCase type name.

    Orval names each file after the type name (e.g. 'sensorSubscriptionResponse.ts'
    exports 'SensorSubscriptionResponse'), so we just uppercase the first character.
    """
    return name[0].upper() + name[1:]


def extract_orval_types(index_path: Path) -> set[str]:
    """Extract exported type names from orval models index.ts.

    Each line looks like: export * from './someTypeName';
    """
    content = index_path.read_text()
    types: set[str] = set()
    for line in content.splitlines():
        m = re.search(r"from\s+['\"]\./([^'\"]+)['\"]", line)
        if m:
            name = m.group(1)
            if name not in ("index",):
                types.add(filename_to_type(name))
    return types


def extract_handwritten_type_names(file_path: Path) -> set[str]:
    """Extract exported type/interface names from a .ts file."""
    content = file_path.read_text()
    names: set[str] = set()
    for m in re.finditer(r"export\s+(interface|type)\s+(\w+)", content):
        kind = m.group(1)
        name = m.group(2)
        if kind == "type" and name in FRONTEND_ONLY_TYPES:
            continue
        names.add(name)
    return names


def find_overlaps(
    handwritten: set[str],
    orval: set[str],
    source_file: str,
) -> list[str]:
    """Find handwritten types that overlap with orval types."""
    problems: list[str] = []
    for name in sorted(handwritten):
        reasons: list[str] = []

        if name in orval:
            reasons.append(f"exact name match: orval/models already exports `{name}`")

        response_name = name + "Response"
        if response_name in orval:
            reasons.append(
                f"suffix match: types.ts `{name}` → orval/models `{response_name}`"
            )

        if reasons:
            problems.append(
                f"  {name} in {source_file}:\n"
                + "\n".join(f"    - {r}" for r in reasons)
            )

    return problems


def main() -> int:
    if not ORVAL_INDEX.exists():
        print(f"Error: orval index not found: {ORVAL_INDEX}", file=sys.stderr)
        return 1

    orval_types = extract_orval_types(ORVAL_INDEX)

    all_problems: list[str] = []

    for file_path, label in [
        (TYPES_FILE, "types.ts"),
        (UI_HELPERS_FILE, "ui-helpers.ts"),
    ]:
        if not file_path.exists():
            print(f"[warn] File not found, skipping: {file_path}", file=sys.stderr)
            continue
        handwritten = extract_handwritten_type_names(file_path)
        problems = find_overlaps(handwritten, orval_types, f"shared/api/{label}")
        all_problems.extend(problems)

    if all_problems:
        print(
            f"[error] {len(all_problems)} type(s) in shared/api types "
            f"already exist in orval/models:\n"
        )
        for p in all_problems:
            print(p)
        return 1

    print("[ok] No duplicate types found between shared/api and orval/models")
    return 0


if __name__ == "__main__":
    sys.exit(main())
